### ISSUE TITLE
Parse from Code

### ISSUE THREAD

**Issue by lucemia** [livingbio/typed-ffmpeg#43]
```

static const AVOption options[] = {
    { "level_in",  "set input gain",     OFFSET(level_in),  AV_OPT_TYPE_DOUBLE, {.dbl=1},        0.015625,   64, A|F|R },
    { "mode",      "set mode",           OFFSET(mode),      AV_OPT_TYPE_INT,    {.i64=0},               0,    1, A|F|R, "mode" },
    {   "downward",0,                    0,                 AV_OPT_TYPE_CONST,  {.i64=0},               0,    0, A|F|R, "mode" },
    {   "upward",  0,                    0,                 AV_OPT_TYPE_CONST,  {.i64=1},               0,    0, A|F|R, "mode" },
    { "threshold", "set threshold",      OFFSET(threshold), AV_OPT_TYPE_DOUBLE, {.dbl=0.125}, 0.000976563,    1, A|F|R },
    { "ratio",     "set ratio",          OFFSET(ratio),     AV_OPT_TYPE_DOUBLE, {.dbl=2},               1,   20, A|F|R },
    { "attack",    "set attack",         OFFSET(attack),    AV_OPT_TYPE_DOUBLE, {.dbl=20},           0.01, 2000, A|F|R },
    { "release",   "set release",        OFFSET(release),   AV_OPT_TYPE_DOUBLE, {.dbl=250},          0.01, 9000, A|F|R },
    { "makeup",    "set make up gain",   OFFSET(makeup),    AV_OPT_TYPE_DOUBLE, {.dbl=1},               1,   64, A|F|R },
    { "knee",      "set knee",           OFFSET(knee),      AV_OPT_TYPE_DOUBLE, {.dbl=2.82843},         1,    8, A|F|R },
    { "link",      "set link type",      OFFSET(link),      AV_OPT_TYPE_INT,    {.i64=0},               0,    1, A|F|R, "link" },
    {   "average", 0,                    0,                 AV_OPT_TYPE_CONST,  {.i64=0},               0,    0, A|F|R, "link" },
    {   "maximum", 0,                    0,                 AV_OPT_TYPE_CONST,  {.i64=1},               0,    0, A|F|R, "link" },
    { "detection", "set detection",      OFFSET(detection), AV_OPT_TYPE_INT,    {.i64=1},               0,    1, A|F|R, "detection" },
    {   "peak",    0,                    0,                 AV_OPT_TYPE_CONST,  {.i64=0},               0,    0, A|F|R, "detection" },
    {   "rms",     0,                    0,                 AV_OPT_TYPE_CONST,  {.i64=1},               0,    0, A|F|R, "detection" },
    { "level_sc",  "set sidechain gain", OFFSET(level_sc),  AV_OPT_TYPE_DOUBLE, {.dbl=1},        0.015625,   64, A|F|R },
    { "mix",       "set mix",            OFFSET(mix),       AV_OPT_TYPE_DOUBLE, {.dbl=1},               0,    1, A|F|R },
    { NULL }
};
```

```
Total filters: 547 / 570
not exists filters {'abuffer', 'abuffersink', 'buffer', 'buffersink'}
not found filters {'azmq', 'convolution_opencl', 'boxblur_opencl', 'vpp_qsv', 'vstack_qsv', 'scale_qsv', 'openclsrc', 'yadif_videotoolbox', 'dilation_opencl', 'coreimagesrc', 'coreimage', 'program_opencl', 'hstack_vaapi', 'xstack_qsv', 'hstack_qsv', 'vstack_vaapi', 'prewitt_opencl', 'roberts_opencl', 'libvmaf_cuda', 'avgblur_opencl', 'xstack_vaapi', 'deinterlace_qsv', 'subtitles', 'zmq', 'sobel_opencl', 'ass', 'erosion_opencl'}
```

- [x] ass
- [x] azmq: 18.34 required zmq
- [x] coreimage: 11.47 required Apple
- [x] coreimagesrc: 16.13
- [ ] libvmaf_cuda: 11.146 required CUDE
- [x] maskedmerge: 11.157
- [x] noise: 11.177
- [x] convolution_opencl
- [x] dilation_opencl
- [x] erosion_opencl
- [x] program_opencl
- [x] boxblur_opencl
- [x] roberts_opencl
- [x] sobel_opencl
- [x] avgblur_opencl
- [x] prewitt_opencl
- [x] openclsrc: 16.11
- [x] xstack_qsv
- [x] vpp_qsv
- [x] hstack_qsv
- [x] deinterlace_qsv
- [x] vstack_qsv
- [x] scale_qsv
- [x] subtitles
- [x] hstack_vaapi
- [x] vstack_vaapi
- [x] xstack_vaapi
- [ ] yadif_videotoolbox (.m)
- [x] xbr: 11.286
- [x] zmq: 18.34 required zmq

**Comment by lucemia** [comment]
```
not exists filters {'buffer', 'GREY_EDGE', 'buffersink'}
not found filters {'openclsrc', 'aeval', 'allyuv', 'agraphmonitor', 'abench', 'frei0r_src', 'xcorrelate', 'adrawgraph', 'bass', 'bandpass', 'highshelf', 'oscilloscope', 'smptebars', 'lut1d', 'vpp_qsv', 'equalizer', 'xstack', 'setrange', 'lowpass', 'thistogram', 'tmix', 'rgbashift', 'perms', 'acue', 'rgbtestsrc', 'asetpts', 'allrgb', 'anlmf', 'allpass', 'setfield', 'pal75bars', 'select', 'deinterlace_qsv', 'haldclutsrc', 'zoneplate', 'bandreject', 'deflate', 'prewitt_opencl', 'afireqsrc', 'tmedian', 'asubcut', 'sharpness_vaapi', 'ainterleave', 'apsnr', 'flip_vulkan', 'sidedata', 'setsar', 'colorhold', 'noformat', 'sobel', 'lutrgb', 'libvmaf_cuda', 'inflate', 'atrim', 'smptehdbars', 'hstack_qsv', 'sidechaingate', 'dilation_opencl', 'xstack_qsv', 'tblend', 'drawgrid', 'lutyuv', 'roberts', 'aintegral', 'loop', 'afifo', 'highpass', 'testsrc2', 'scale2ref_npp', 'asettb', 'yuvtestsrc', 'asisdr', 'testsrc', 'adeclip', 'asendcmd', 'showspectrumpic', 'tiltshelf', 'lowshelf', 'vstack', 'azmq', 'hsvhold', 'vstack_qsv', 'acompressor', 'alphaextract', 'arealtime', 'haldclut', 'greyedge', 'deconvolve', 'yadif_videotoolbox', 'chromahold', 'colorspectrum', 'kirsch', 'biquad', 'scale_qsv', 'prewitt', 'boxblur_opencl', 'pal100bars', 'areverse', 'lut', 'interlace', 'coreimage', 'maskedmax', 'asplit', 'nullsrc', 'dilation', 'subtitles', 'amovie', 'colorchart', 'treble', 'asuperpass', 'pixscope', 'asegment', 'alatency', 'erosion', 'msad', 'roberts_opencl', 'tlut2', 'acrossfade', 'metadata', 'vstack_vaapi', 'hstack_vaapi', 'astreamselect', 'scale2ref', 'doubleweave', 'showwavespic', 'sobel_opencl', 'vflip_vulkan', 'xstack_vaapi', 'unpremultiply', 'scharr', 'coreimagesrc', 'aphaseshift', 'asuperstop'}
```

**Comment by lucemia** [comment]
parse ffmpeg directly
```
Total filters: 440 / 570
not exists filters {'buffersink', 'GREY_EDGE', 'buffer'}
not found filters {'prewitt_opencl', 'asplit', 'haldclutsrc', 'pal100bars', 'libvmaf_cuda', 'sidedata', 'tiltshelf', 'allyuv', 'smptebars', 'afireqsrc', 'coreimagesrc', 'areverse', 'tmedian', 'msad', 'maskedmax', 'aintegral', 'equalizer', 'sobel_opencl', 'rgbtestsrc', 'acue', 'abench', 'hstack_vaapi', 'greyedge', 'azmq', 'aeval', 'ainterleave', 'loop', 'colorchart', 'openclsrc', 'lutyuv', 'xstack_qsv', 'hsvhold', 'sobel', 'acompressor', 'flip_vulkan', 'allpass', 'asendcmd', 'pal75bars', 'deflate', 'vpp_qsv', 'vstack_vaapi', 'prewitt', 'select', 'interlace', 'highshelf', 'erosion', 'scale_qsv', 'testsrc', 'zoneplate', 'drawgrid', 'setsar', 'scale2ref', 'apsnr', 'arealtime', 'bandpass', 'dilation_opencl', 'coreimage', 'adeclip', 'hstack_qsv', 'lowpass', 'testsrc2', 'tmix', 'oscilloscope', 'inflate', 'haldclut', 'doubleweave', 'asetpts', 'deinterlace_qsv', 'afifo', 'boxblur_opencl', 'asubcut', 'asuperstop', 'bass', 'xstack_vaapi', 'metadata', 'xcorrelate', 'lut', 'acrossfade', 'lowshelf', 'thistogram', 'colorspectrum', 'perms', 'roberts', 'subtitles', 'bandreject', 'noformat', 'setrange', 'yuvtestsrc', 'lut1d', 'asettb', 'showspectrumpic', 'smptehdbars', 'kirsch', 'scale2ref_npp', 'alatency', 'aphaseshift', 'vstack', 'astreamselect', 'asegment', 'unpremultiply', 'highpass', 'biquad', 'chromahold', 'anlmf', 'lutrgb', 'allrgb', 'rgbashift', 'vflip_vulkan', 'treble', 'asuperpass', 'atrim', 'amovie', 'pixscope', 'asisdr', 'colorhold', 'sidechaingate', 'yadif_videotoolbox', 'tblend', 'nullsrc', 'alphaextract', 'dilation', 'adrawgraph', 'sharpness_vaapi', 'showwavespic', 'deconvolve', 'scharr', 'agraphmonitor', 'setfield', 'tlut2', 'roberts_opencl', 'frei0r_src', 'xstack', 'vstack_qsv'}
```

parse compiled ffmpeg

```
Total filters: 386 / 570
not exists filters {'buffer', 'buffersink'}
not found filters {'pal100bars', 'lutrgb', 'ass', 'asplit', 'bass', 'bwdif_cuda', 'haldclut', 'flite', 'bandpass', 'kirsch', 'scale_vulkan', 'atrim', 'zscale', 'chromahold', 'colorchart', 'vstack_vaapi', 'procamp_vaapi', 'allyuv', 'noformat', 'drawgrid', 'libplacebo', 'highshelf', 'treble', 'rgbtestsrc', 'yuvtestsrc', 'ddagrab', 'sofalizer', 'acue', 'smptebars', 'lensfun', 'tmedian', 'setsar', 'ainterleave', 'convolution_opencl', 'tonemap_vaapi', 'showspectrumpic', 'vflip_vulkan', 'sidedata', 'ladspa', 'smptehdbars', 'prewitt_opencl', 'xstack_qsv', 'avgblur_opencl', 'roberts_opencl', 'scale_vaapi', 'erosion_opencl', 'afifo', 'asettb', 'areverse', 'blend_vulkan', 'iccgen', 'overlay_cuda', 'program_opencl', 'asubcut', 'interlace', 'deinterlace_qsv', 'yadif_cuda', 'drawtext', 'vidstabtransform', 'subtitles', 'tmix', 'vidstabdetect', 'amovie', 'xfade_vulkan', 'aeval', 'nlmeans_vulkan', 'lut1d', 'doubleweave', 'libvmaf_cuda', 'alphaextract', 'scale_npp', 'zmq', 'afireqsrc', 'allpass', 'asr', 'bandreject', 'deinterlace_vaapi', 'bs2b', 'agraphmonitor', 'sharpness_vaapi', 'coreimage', 'bwdif_vulkan', 'aphaseshift', 'scale_cuda', 'loop', 'anlmf', 'hstack_qsv', 'lutyuv', 'aintegral', 'abench', 'asisdr', 'deflate', 'denoise_vaapi', 'setfield', 'nullsrc', 'acompressor', 'select', 'sobel', 'frei0r_src', 'oscilloscope', 'xcorrelate', 'flip_vulkan', 'boxblur_opencl', 'hflip_vulkan', 'avgblur_vulkan', 'biquad', 'dilation', 'deconvolve', 'frei0r', 'haldclutsrc', 'thistogram', 'setrange', 'scale_qsv', 'ocr', 'unpremultiply', 'iccdetect', 'rgbashift', 'colorspace_cuda', 'lowpass', 'color_vulkan', 'coreimagesrc', 'pixscope', 'prewitt', 'colorhold', 'rubberband', 'scale2ref_npp', 'acrossfade', 'openclsrc', 'lowshelf', 'tiltshelf', 'chromakey_cuda', 'vpp_qsv', 'colorspectrum', 'thumbnail_cuda', 'tlut2', 'adrawgraph', 'transpose_vaapi', 'roberts', 'alatency', 'hsvhold', 'xstack', 'ocv', 'perms', 'sidechaingate', 'pal75bars', 'scharr', 'adeclip', 'azmq', 'gblur_vulkan', 'tblend', 'bilateral_cuda', 'yadif_videotoolbox', 'inflate', 'overlay_qsv', 'arealtime', 'testsrc', 'sharpen_npp', 'vstack_qsv', 'allrgb', 'asuperpass', 'msad', 'apsnr', 'asetpts', 'asegment', 'astreamselect', 'transpose_npp', 'maskedmax', 'scale2ref', 'hstack_vaapi', 'transpose_vulkan', 'testsrc2', 'lv2', 'asuperstop', 'overlay_vaapi', 'libvmaf', 'overlay_vulkan', 'asendcmd', 'showwavespic', 'highpass', 'xstack_vaapi', 'zoneplate', 'metadata', 'chromaber_vulkan', 'sobel_opencl', 'vstack', 'dilation_opencl'}
```

**Comment by lucemia** [comment]
```
Total filters: 437 / 570
not exists filters {'buffersink', 'buffer'}
not found filters {'pixscope', 'xstack', 'zoneplate', 'asisdr', 'prewitt', 'openclsrc', 'lutyuv', 'asuperpass', 'tmix', 'tiltshelf', 'vflip_vulkan', 'deflate', 'select', 'anlmf', 'boxblur_opencl', 'roberts_opencl', 'scale_qsv', 'testsrc', 'agraphmonitor', 'xstack_qsv', 'adeclip', 'frei0r_src', 'asplit', 'haldclut', 'haldclutsrc', 'thistogram', 'prewitt_opencl', 'asettb', 'bandpass', 'chromahold', 'vstack_vaapi', 'xstack_vaapi', 'nullsrc', 'highpass', 'allrgb', 'abench', 'roberts', 'yuvtestsrc', 'rgbtestsrc', 'lutrgb', 'aeval', 'vstack', 'sobel', 'hstack_vaapi', 'arealtime', 'smptehdbars', 'setsar', 'allyuv', 'bass', 'lut1d', 'scale2ref', 'drawgrid', 'asendcmd', 'libvmaf_cuda', 'flip_vulkan', 'noformat', 'hstack_qsv', 'astreamselect', 'deconvolve', 'oscilloscope', 'asegment', 'showspectrumpic', 'coreimagesrc', 'biquad', 'msad', 'aintegral', 'allpass', 'erosion_opencl', 'asuperstop', 'rgbashift', 'smptebars', 'pal100bars', 'asetpts', 'aphaseshift', 'highshelf', 'doubleweave', 'xcorrelate', 'setfield', 'areverse', 'lowpass', 'showwavespic', 'tmedian', 'vstack_qsv', 'apsnr', 'tblend', 'deinterlace_qsv', 'colorhold', 'afireqsrc', 'colorchart', 'testsrc2', 'coreimage', 'convolution_opencl', 'zmq', 'maskedmax', 'avgblur_opencl', 'colorspectrum', 'afifo', 'hsvhold', 'unpremultiply', 'atrim', 'acue', 'ass', 'metadata', 'kirsch', 'treble', 'bandreject', 'loop', 'alatency', 'scharr', 'perms', 'setrange', 'tlut2', 'pal75bars', 'amovie', 'yadif_videotoolbox', 'acompressor', 'interlace', 'sobel_opencl', 'subtitles', 'inflate', 'scale2ref_npp', 'sharpness_vaapi', 'lowshelf', 'acrossfade', 'dilation', 'dilation_opencl', 'adrawgraph', 'ainterleave', 'sidechaingate', 'alphaextract', 'vpp_qsv', 'azmq', 'sidedata', 'asubcut', 'program_opencl'}
```

ignore compile error

**Comment by lucemia** [comment]
```
Total filters: 544 / 570
not exists filters {'buffer', 'buffersink', 'abuffer', 'abuffersink'}
not found filters {'hstack_vaapi', 'roberts_opencl', 'sobel_opencl', 'hstack_qsv', 'erosion_opencl', 'xstack_vaapi', 'coreimage', 'convolution_opencl', 'maskedmerge', 'vstack_vaapi', 'vstack_qsv', 'deinterlace_qsv', 'subtitles', 'openclsrc', 'coreimagesrc', 'zmq', 'avgblur_opencl', 'scale_qsv', 'noise', 'xbr', 'vpp_qsv', 'libvmaf_cuda', 'prewitt_opencl', 'dilation_opencl', 'boxblur_opencl', 'yadif_videotoolbox', 'azmq', 'xstack_qsv', 'ass', 'program_opencl'}
```

**Comment by lucemia** [comment]
In the FFmpeg project, files that end with the `.m` extension are typically Objective-C source files. Objective-C is a programming language used primarily on Apple's macOS and iOS operating systems. It's an extension of the C language and is used for developing applications within Apple's ecosystem.

Here are a few key points about `.m` files in FFmpeg:

1. **Objective-C Integration:** The `.m` files in FFmpeg are likely there to provide integration with macOS and iOS platforms. This could be for features that are specific to these platforms, such as hardware acceleration, using Apple's APIs for video and audio processing, or interfacing with system components like the camera or microphone on Apple devices.

2. **Cross-Platform Compatibility:** FFmpeg is a cross-platform project, meaning it runs on a variety of operating systems, including Windows, Linux, and macOS. The use of Objective-C files is part of this cross-platform strategy, catering specifically to Apple's platforms.

3. **Conditional Compilation:** These Objective-C files are likely compiled only when FFmpeg is being built for macOS or iOS. Conditional compilation directives (like `#ifdef` macros) are used to ensure that these files are only included in the build process when necessary.

4. **Hardware Acceleration and API Access:** Apple provides various APIs and hardware acceleration features that are accessible through Objective-C. FFmpeg may use these APIs to enhance performance or compatibility with Apple hardware.

In summary, the `.m` files in FFmpeg are for code that is written in Objective-C, targeting functionalities specific to Apple's operating systems. This allows FFmpeg to utilize platform-specific features and integrate more seamlessly with macOS and iOS environments.

**Comment by lucemia** [comment]
[13 VAAPI Video Filters](https://ffmpeg.org/ffmpeg-filters.html#toc-VAAPI-Video-Filters)
VAAPI Video filters are usually used with VAAPI decoder and VAAPI encoder. Below is a description of VAAPI video filters.

To enable compilation of these filters you need to configure FFmpeg with --enable-vaapi.

To use vaapi filters, you need to setup the vaapi device correctly. For more information, please read https://trac.ffmpeg.org/wiki/Hardware/VAAPI

不是傳統的 Video, Audio, MultiMedia Filter, 先不處理

**Comment by lucemia** [comment]
[12 OpenCL Video Filters](https://ffmpeg.org/ffmpeg-filters.html#toc-OpenCL-Video-Filters)
Below is a description of the currently available OpenCL video filters.

To enable compilation of these filters you need to configure FFmpeg with --enable-opencl.

Running OpenCL filters requires you to initialize a hardware device and to pass that device to all filters in any filter graph.

-init_hw_device opencl[=name][:device[,key=value...]]
Initialise a new hardware device of type opencl called name, using the given device parameters.

-filter_hw_device name
Pass the hardware device called name to all filters in any filter graph.

For more detailed information see https://www.ffmpeg.org/ffmpeg.html#Advanced-Video-options

Example of choosing the first device on the second platform and running avgblur_opencl filter with default parameters on it.
-init_hw_device opencl=gpu:1.0 -filter_hw_device gpu -i INPUT -vf "hwupload, avgblur_opencl, hwdownload" OUTPUT
Since OpenCL filters are not able to access frame data in normal memory, all frame data needs to be uploaded([hwupload](https://ffmpeg.org/ffmpeg-filters.html#hwupload)) to hardware surfaces connected to the appropriate device before being used and then downloaded([hwdownload](https://ffmpeg.org/ffmpeg-filters.html#hwdownload)) back to normal memory. Note that [hwupload](https://ffmpeg.org/ffmpeg-filters.html#hwupload) will upload to a surface with the same layout as the software frame, so it may be necessary to add a [format](https://ffmpeg.org/ffmpeg-filters.html#format) filter immediately before to get the input into the right format and [hwdownload](https://ffmpeg.org/ffmpeg-filters.html#hwdownload) does not support all [format](https://ffmpeg.org/ffmpeg-filters.html#format)s on the output - it may be necessary to insert an additional format filter immediately following in the graph to get the output in a supported format.

**Comment by lucemia** [comment]
[15 QSV Video Filters](https://ffmpeg.org/ffmpeg-filters.html#toc-QSV-Video-Filters)
Below is a description of the currently available QSV video filters.

To enable compilation of these filters you need to configure FFmpeg with --enable-libmfx or --enable-libvpl.

To use QSV filters, you need to setup the QSV device correctly. For more information, please read https://trac.ffmpeg.org/wiki/Hardware/QuickSync

**Comment by lucemia** [comment]
ass, subtitle 由於 MACRO 的關西，直接 copy 才能用

**Comment by lucemia** [comment]
cuda 也是一樣

**Comment by lucemia** [comment]
The error "'ass/ass.h' file not found" in macOS typically occurs when you're trying to compile a program that depends on the libass library, but the library is not installed or not properly linked. libass (lib Advanced SubStation Alpha) is a portable subtitle renderer for the ASS/SSA (Advanced SubStation Alpha/SubStation Alpha) subtitle format.

Here are the steps to resolve this issue:

1. **Install libass:**
   The easiest way to install libass on macOS is via Homebrew, a popular package manager for macOS. If you don't have Homebrew installed, you can install it from [https://brew.sh/](https://brew.sh/). Then, use Homebrew to install libass:

   ```bash
   brew install libass
   ```

2. **Linking the Library:**
   When compiling your program, you need to ensure that the compiler knows where to find the libass headers and libraries. You can use the `-I` flag to specify the include path and `-L` flag to specify the library path. Typically, for Homebrew installations, they are located in `/usr/local/include` and `/usr/local/lib`, respectively.

   For example:
   ```bash
   gcc -o your_program your_program.c -I/usr/local/include -L/usr/local/lib -lass
   ```

3. **Using pkg-config:**
   If `libass` is installed with `pkg-config` support, you can use `pkg-config` to automatically set the correct flags. First, check if `libass` is available via `pkg-config`:

   ```bash
   pkg-config --libs --cflags libass
   ```

   If this command returns the necessary compiler and linker flags, you can use them directly in your compile command:

   ```bash
   gcc -o your_program your_program.c $(pkg-config --libs --cflags libass)
   ```

4. **Environment Variables:**
   Sometimes, setting the `PKG_CONFIG_PATH` environment variable is necessary so that `pkg-config` can find the metadata files for libass:

   ```bash
   export PKG_CONFIG_PATH="/usr/local/lib/pkgconfig"
   ```

5. **Check Your Code:**
   Ensure that your code includes the libass header correctly, usually something like:

   ```c
   #include <ass/ass.h>
   ```

6. **Reconfigure and Rebuild:**
   If you're working with a pre-existing build system (like with FFmpeg or other projects that use libass), you may need to re-run the `configure` script or equivalent to pick up the newly installed library.

7. **Verify Installation:**
   After installation, you can verify if libass is correctly installed and linked by compiling a small test program that includes `ass/ass.h`.

By following these steps, you should be able to resolve the "'ass/ass.h' file not found" error on macOS.
